File: tests/Parity/declarations/oracle.py
#!/usr/bin/env python3
"""Oracle: chạy DeclarationTracker của compiler Python.

stdin  : mỗi dòng một đường dẫn .sao
stdout : <đường dẫn><TAB><JSON danh sách khai báo>
"""
import contextlib
import json
import os
import sys


def _repo_root() -> str:
    path = os.path.abspath(__file__)
    while path != os.path.dirname(path):
        path = os.path.dirname(path)
        if os.path.isdir(os.path.join(path, 'builder', '.reference', 'python', 'src')):
            return path
    raise RuntimeError('Không tìm thấy repo root')


ROOT = _repo_root()
sys.path.insert(0, os.path.join(ROOT, 'builder', '.reference', 'python', 'src'))

from common.declaration_tracker import DeclarationTracker  # noqa: E402


def main() -> int:
    rows = []

    with contextlib.redirect_stdout(sys.stderr):
        for line in sys.stdin:
            file = line.strip()
            if not file:
                continue

            try:
                with open(file, encoding='utf-8') as f:
                    content = f.read()
                # Instance MỚI mỗi file — tracker giữ trạng thái giữa các lần gọi
                decls = DeclarationTracker().parse_all_declarations(content)
                # Bỏ 'position' — xem giải thích trong run.sh
                payload = [{k: v for k, v in d.items() if k != 'position'} for d in decls]
            except Exception as exc:  # noqa: BLE001
                payload = {'__error__': type(exc).__name__}

            rows.append((os.path.relpath(file, ROOT),
                         json.dumps(payload, ensure_ascii=False, separators=(',', ':'))))

    for rel, payload in rows:
        print(f'{rel}\t{payload}')

    return 0


if __name__ == '__main__':
    sys.exit(main())

File: tests/Parity/source-split/oracle.js
#!/usr/bin/env node
/**
 * Oracle: tách file .sao bằng CHÍNH parseSaoFile của Saola Builder.
 *
 * Khác các cổng khác, oracle ở đây là JavaScript chứ không phải Python —
 * parseSaoFile nằm ở phía Node (builder/src/index.js). Xem
 * docs/01-architecture.md §3 về việc vì sao phần này cũng phải port.
 *
 * stdin  : mỗi dòng một đường dẫn file .sao
 * stdout : <đường dẫn><TAB><JSON các phần tách được>
 */
'use strict';

const fs = require('fs');
const path = require('path');

function repoRoot() {
    let dir = __dirname;
    while (dir !== path.dirname(dir)) {
        dir = path.dirname(dir);
        if (fs.existsSync(path.join(dir, 'builder', 'src', 'index.js'))) return dir;
    }
    throw new Error('Không tìm thấy repo root (thư mục chứa builder/src/index.js)');
}

const ROOT = repoRoot();
const Compiler = require(path.join(ROOT, 'builder', 'src', 'index.js'));
const compiler = new Compiler();

// `ssrContent` không được port: luôn là '' và không nơi nào đọc.
const FIELDS = ['declarations', 'blade', 'bladeWithSSR', 'script', 'style', 'cleanedContent', 'wrapperType'];

const input = fs.readFileSync(0, 'utf-8');

for (const line of input.split('\n')) {
    const file = line.trim();
    if (!file) continue;

    let payload;
    try {
        const parts = compiler.parseSaoFile(fs.readFileSync(file, 'utf-8'), file);
        payload = {};
        for (const key of FIELDS) {
            payload[key] = parts[key] === undefined ? null : parts[key];
        }
    } catch (error) {
        payload = { __error__: error.constructor.name };
    }

    process.stdout.write(`${path.relative(ROOT, file)}\t${JSON.stringify(payload)}\n`);
}
